Prepare for release 0.1.0 (#16)
software-mansion/cairo-annotations: README.md

# Cairo Annotations

The `cairo-annotations` crate provides tools tailored for working with annotations in the Cairo language.

> These
> annotations are part of the Sierra Debug Information format and serve to enrich the code by detailing information.

Reviewer: ```suggestion
These annotations are part of the Sierra Debug Information format and serve to enrich the code by providing detailed information.
```

> 📝 **Note**
>
> Although this crate is primarily utilized by projects like `cairo-coverage`, `cairo-profiler`, and `starknet-foundry`,

> > it is also fully capable of functioning as a standalone library to work with Sierra Debug Information annotations.

Reviewer: do we want to give an example on that?

Author: I think `Profiler Annotations` and `Coverage Annotations` sections in README provide examples of that. Or do you mean something else?

Reviewer: I guess, but the examples are leaning heavily towards these tools. I think it may not be that obvious for some what this note really mean. Maybe it is not that relevant alltogether?

Author: To be honest after some consideration, I believe this library use outside of these tools should not be recommended, as the API is primarily designed for them.

What about this note?

> This library is primarily used by projects such as `cairo-coverage`, `cairo-profiler`, and `starknet-foundry`, and is specifically tailored to meet their needs. Consequently, it may not be suitable for use outside of these tools.

Reviewer: Maybe rather something like "not recommended" or just "use at your own risk"? I don't think we should actively discourage people, but it should be pointed out that things may not work out smoothly for everyone, so they should not expect things to "just work". Coming back to this: `Maybe it (the note) is not that relevant alltogether?` perhaps this would be the best way, to just not write this at all and only state that "it is what it is, and may not be for you if you do not know what you're doing" (but sugarcoated a bit 😅 )?

Author: I think you are right that the note is not relevant, so I will delete this. Discouraging people in this situation isn't going to do anything good for us.

## Features of Cairo Annotations

### Structured Annotations

`cairo-annotations` offers a structured representation, allowing for more ergonomic manipulation of annotations.

> Some
> key features include:

Reviewer: ```suggestion
Some key features include:
```

I can see there are multiple places having a newline whitespace after first word. Is this something intended?

Author: No, it seems like a mistake with formatter 😅. I will fix it

- **Coverage Annotations**: Track locations in the Cairo code that correspond to specific Sierra statements.
- **Profiler Annotations**: Provide mappings from Sierra statements to fully qualified Cairo paths, detailing which
  functions in the Cairo code triggered them.

All annotations implement the `TryFromDebugInfo` trait, enabling their extraction from Sierra debug information.
Here's
a simple example:

```rust

> let annotations = VersionedCoverageAnnotations::try_from_debug_info(sierra_debug_info).unwrap();

Reviewer: it may be a good idea to show what `annotation` will look like? not sure though

Author: You mean json form?

Reviewer: yes, but probably with heavily mocked data. or at least some indication on return type or something

Author: Added

Reviewer: perfect, thanks. Hopefully this won't be a pain to maintain 🙏

```

### Coverage Annotations

Coverage annotations provide a mapping from Sierra statement indices to sources in the Cairo code that resulted in their
creation.
For extensive documentation,

> see [CoverageAnnotationsV1](./crates/cairo-annotations/src/annotations/coverage.rs).

Reviewer: nit: for now it does not matter much, but for the future it would be nice to have an information about versioning * somewhere *.

Author: Will add this 🔜

Example to retrieve code location information:

```rust
use cairo_annotations::annotations::coverage::{
    CodeLocation, ColumnNumber, CoverageAnnotationsV1, LineNumber, SourceCodeLocation,
    SourceCodeSpan, SourceFileFullPath, VersionedCoverageAnnotations,
};
use cairo_annotations::annotations::TryFromDebugInfo;
use cairo_lang_sierra::program::StatementIdx;

let VersionedCoverageAnnotations::V1(annotations) =

> VersionedCoverageAnnotations::try_from_debug_info(sierra_debug_info).unwrap();

Reviewer: ```suggestion
    VersionedCoverageAnnotations::try_from_debug_info(sierra_debug_info).unwrap();
```
nit. let's properly format the code snippets, so they're easier on the eyes :)

.unwrap();

assert_eq!(
    code_locations,
    &[CodeLocation(
        SourceFileFullPath("/path/to/my/file.cairo".into()),
        SourceCodeSpan {
            start: SourceCodeLocation { line: LineNumber(7), col: ColumnNumber(4) },
            end: SourceCodeLocation { line: LineNumber(7), col: ColumnNumber(4) },
        }
    )]
);
```

### Profiler Annotations

Profiler annotations map Sierra statement indices to Cairo function paths, showing which functions led to their
generation.
Detailed information is available
in [ProfilerAnnotationsV1](./crates/cairo-annotations/src/annotations/profiler.rs).

Example to obtain the Cairo path:

```rust
use cairo_annotations::annotations::profiler::{
    FunctionName, ProfilerAnnotationsV1, VersionedProfilerAnnotations,
};
use cairo_annotations::annotations::TryFromDebugInfo;
use cairo_lang_sierra::program::StatementIdx;

let VersionedProfilerAnnotations::V1(annotations) =
VersionedProfilerAnnotations::try_from_debug_info(sierra_debug_info).unwrap();

let functions_names = annotations
.statements_functions
.get( & StatementIdx(331))
.unwrap();

assert_eq!(
    functions_names,
    &[FunctionName("scarb_template::fib".into())]
);
```

> ## Integration with Snforge

Reviewer: ```suggestion
## Integration with snforge
```
nit: There was a discussion around capital letters/camel cases etc some time ago. The consensus was to just use small letters for the naming, I believe this is also a case for snforge/sncast in the foundry docs :+1:
